### PR TITLE
AWS tutorial

### LINKED ISSUE #1109: Component graph is missing
https://axom.readthedocs.io/en/develop/#component-level-dependencies

There used to be a picture showing the dot graph. It is apparently broken.

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This PR adds an axom tutorial building up to a GPU- and spatial index-accelerated application akin to our `mesh_tester` utility (in `<axom>/src/tools/mesh_tester`)
- The tutorial currently has lessons on 
   - building/configuring an application against an installed axom
   - reading in an STL mesh
   - developing a naive self-intersection algorithm
- The following lessons will be added in short order
   - [x] extending the naive algorithm using RAJA and Umpire via `axom::forall`
   - [x] developing a RAJA-based algorithm using spatial indexes
 - [x] I will also add a top-level `README` outlining the structure of the tutorial

Misc changes:
- Fixes some warnings in Release configs
- Optimizes `primal::intersect(BoundingBox, BoundingBox)`
- Fixes a bug with `axom::utilities::locale()` when the user's default locale is missing
- Fixes #1109 -- we can once again generate graphs in our readthedocs builds

### CODE REVIEW

File: src/examples/radiuss_tutorial/lesson_00/README.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+Our application can be configured and built using the following commands:
+```shell
+> cd <path/to/tutorial/files>
+> mkdir build
+> cd build
+> cmake -C ../host-config.cmake ..
+> make -j16
+```
+
+!!! tip BLT provides a few lightweight "smoke" tests to help diagnose configuration issues. If we configure our application with the ``ENABLE_TESTS`` CMake option, the above command will build these tests and we can run them using ``ctest``.
```
Comment (kennyweiss) [inline src/examples/radiuss_tutorial/lesson_00/README.md:32]: These ``!!! tip`` annotations look good in my VS-code environment. I will update them to github markdown after the tutorial. The latter has the form

\> \*\*note\*\*
\> A note here.

e.g.


> **note**
> A note here.

**Review by rhornung67 — APPROVED** [review]

File: src/axom/primal/geometry/BoundingBox.hpp [hunk 2]
```diff
@@ -436,18 +436,19 @@ template <typename OtherType>
...
-
   // AABBs cannot intersect if they are separated along any dimension
   for(int i = 0; i < NDIMS; ++i)
   {
-    status &= detail::intersect_bbox_bbox(m_min[i],
-                                          m_max[i],
-                                          otherBB.m_min[i],
-                                          otherBB.m_max[i]);
-  }  // END for all dimensions
+    if(!detail::intersect_bbox_bbox(m_min[i],
```
Comment (white238) [inline src/axom/primal/geometry/BoundingBox.hpp:442]: 👍 

Nice optimization

File: src/axom/primal/operators/detail/intersect_bounding_box_impl.hpp [hunk 3]
```diff
@@ -37,8 +37,7 @@ AXOM_HOST_DEVICE inline bool intersect_bbox_bbox(const T& min1,
                                                  const T& min2,
                                                  const T& max2)
 {
-  bool status = ((max1 < min2 || min1 > max2) ? false : true);
-  return status;
+  return max1 >= min2 && min1 <= max2;
```
Comment (white238) [inline src/axom/primal/operators/detail/intersect_bounding_box_impl.hpp:40]: I cant make a comment far away from this but can you fix `drection` on line 23 and 25?

File: src/examples/CMakeLists.txt [hunk 4]
```diff
@@ -66,3 +66,28 @@ install(
         examples/axom/using-with-blt
 )
 
+
+#------------------------------------------------------------------------------
+# configure and install 'radiuss_tutorial' example
+#------------------------------------------------------------------------------
+configure_file(
+    radiuss_tutorial/host-config.cmake.in
```
Comment (white238) [inline src/examples/CMakeLists.txt:74]: I think I would just call this the tutorial... but we can debate/fix that later.

File: src/examples/radiuss_tutorial/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+find_dependency(axom REQUIRED
+                NO_DEFAULT_PATH 
+                PATHS ${AXOM_DIR}/lib/cmake)
+
+#------------------------------------------------------------------------------
+# Set up target executables for the tutorial
+#------------------------------------------------------------------------------
+
+set(tutorial_deps axom axom::fmt axom::cli11)
+blt_list_append(TO tutorial_deps ELEMENTS cuda IF CUDA_FOUND)
```
Comment (white238) [inline src/examples/radiuss_tutorial/CMakeLists.txt:56]:
```suggestion
blt_list_append(TO tutorial_deps ELEMENTS cuda IF ENABLE_CUDA)
```

File: src/examples/radiuss_tutorial/host-config.cmake.in [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+# SPDX-License-Identifier: (BSD-3-Clause)
+#------------------------------------------------------------------------------
+# Host-config for 'radiuss_tutorial' example. 
+# Defines configuration variables from Axom's build.
+#------------------------------------------------------------------------------
+
+# Config related to compiler
+set(CMAKE_C_COMPILER              "@CMAKE_C_COMPILER@"       CACHE PATH "")
+set(CMAKE_CXX_COMPILER            "@CMAKE_CXX_COMPILER@"     CACHE PATH "")
+set(CMAKE_Fortran_COMPILER        "@CMAKE_Fortran_COMPILER@" CACHE PATH "")
```
Comment (white238) [inline src/examples/radiuss_tutorial/host-config.cmake.in:13]: Is fortran being used in this tutorial?

File: src/axom/primal/geometry/BoundingBox.hpp (hunk 2 above)
Comment (kennyweiss) [reply src/axom/primal/geometry/BoundingBox.hpp:442]: Thanks! For the "naive" self-intersection algorithm in the tutorial, the `intersect(BoundingBox,BoundingBox)` test used something like half the cycles on a test that ran for a few seconds. This optimization improved the runtime by something like 30% on our blueos Power9s.

I still need to benchmark this before merging to ensure that it didn't give up performance on Intel runs.

Tag: @publixsubfan

File: src/examples/radiuss_tutorial/lesson_00/README.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+ { core; inlet; klee; lumberjack; mint; primal; quest; sidre; slam; slic; spin; }
+Active Dependencies: 
+ { CLI11; conduit; fmt; hdf5; lua; mfem; mpi; sol; sparsehash; raja; umpire; }
+```
+
+> :clapper: Run the example for this lesson
+
+
+### Next time:
+In the next lesson, we will add a logger and command line parser and load in a triangle mesh
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_00/README.md:139]:
```suggestion
In the next lesson, we will add a logger and command line parser and load in a triangle mesh. 
```

File: src/examples/radiuss_tutorial/lesson_00/check_axom_configuration.cpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#include <iostream>
+
+int main()
+{
+  std::cout << "Checking properties of installed Axom library. \n";
+
+  // Check Axom version
+  std::cout << axom::fmt::format("Version: {}", axom::getVersion()) << "\n\n";
+
+  // Print axom about
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_00/check_axom_configuration.cpp:25]:
```suggestion
  // Print Axom about
```

File: src/examples/radiuss_tutorial/lesson_01/README.md [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+* ``quest`` -- has a utility function for loading the STL mesh
+* ``mint`` -- provides a class for triangle meshes
+
+## Using a SLIC-based logger
+
+We begin by setting up a logger for our application using axom's `slic` component ([Simple Logging Interface Code](https://axom.readthedocs.io/en/develop/axom/slic/docs/sphinx/index.html)).
+
+`slic` uses a C-style API over a singleton instance and also provides a few basic logging macros that applications can either use or extend/build upon. 
+
+For our application, we will use an RAII-based wrapper struct
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_01/README.md:19]:
```suggestion
For our application, we will use an RAII-based wrapper struct:
```

File: src/examples/radiuss_tutorial/lesson_01/README.md [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+
+This example uses the following Axom components:
+* ``core`` -- has a dynamic array class ``axom::Array``
+* ``slic`` -- provides logging functionality
+* ``quest`` -- has a utility function for loading the STL mesh
+* ``mint`` -- provides a class for triangle meshes
+
+## Using a SLIC-based logger
+
+We begin by setting up a logger for our application using axom's `slic` component ([Simple Logging Interface Code](https://axom.readthedocs.io/en/develop/axom/slic/docs/sphinx/index.html)).
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_01/README.md:15]:
```suggestion
We begin by setting up a logger for our application using Axom's `slic` component ([Simple Logging Interface Code](https://axom.readthedocs.io/en/develop/axom/slic/docs/sphinx/index.html)).
```

File: src/examples/radiuss_tutorial/lesson_01/README.md [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+      new slic::GenericOutputStream(&std::cout, slicFormatStr);
+    slic::addStreamToMsgLevel(compactStream, slic::message::Info);
+    slic::addStreamToMsgLevel(compactStream, slic::message::Debug);
+  }
+
+  ~BasicLogger() { axom::slic::finalize(); }
+};
+```
+
+The constructor calls ``slic::initialize()``, sets the logging level to ``slic::message::Debug`` and then creates log streams for the different message levels. The destructor finalizes the slic component with `slic::finalize()`.
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_01/README.md:51]:
```suggestion
The constructor calls ``slic::initialize()``, sets the logging level to ``slic::message::Debug`` and then creates log streams for the different message levels. The destructor finalizes the SLIC component with `slic::finalize()`.
```

Comment (white238) [reply src/examples/radiuss_tutorial/lesson_01/README.md:51]: Is component the right word here in the last sentence?

File: src/examples/radiuss_tutorial/lesson_02/README.md [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+## Primal: Computational geometry primitives and operations
+Axom's ``primal`` component provides robust computational geometry primitives as well as operators between pairs of primitives.
+In this lesson, we will be using the following primitives, which are defined as templated classes:
+* ``primal::Point`` --  represents the coordinates of a point in space
+* ``primal::Triangle`` -- represents a triangle as a triplet of ``primal::Point``s
+* ``primal::BoundingBox`` -- represents an axis-aligned slab bounding a region of space
+
+Since computational geometry primitives are typically used in performance-critical code sections, ``primal``'s primitives and operators are typically templated on the spatial dimension as well as the coordinate types. Thus, e.g., we have ``primal::Point<double, 3>`` for a point with 64-bit floating point coordinates in three-dimensional space. 
+
+To simplify code readability, we often use type aliases when the types are known. For example, if our application always uses 3D triangles with floating point coordinate types, we might define
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_02/README.md:112]:
```suggestion
To simplify code readability, we often use type aliases when the types are known. For example, if our application always uses 3D triangles with floating point coordinate types, we might define:
```

File: src/examples/radiuss_tutorial/lesson_02/README.md [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+      {
+        intersectionPairs.emplace_back(std::make_pair(idx1, idx2));
+      }
+    }
+  }
+
+  return intersectionPairs;
+}
+```
+where the ``trianglesIntersect`` lambda is defined as:
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_02/README.md:188]:
```suggestion
Where the ``trianglesIntersect`` lambda is defined as:
```

File: src/axom/core/utilities/System.cpp [hunk 14]
```diff
@@ -97,7 +97,7 @@ std::locale locale(const std::string& name)
   }
   catch(std::runtime_error&)
   {
-    loc = std::locale(loc, "", std::locale::ctype);
+    loc = std::locale(loc, "C", std::locale::ctype);
```
Comment (kennyweiss) [inline src/axom/core/utilities/System.cpp:100]: Locales are the gift that keep on giving!

The standard appears to require `std::locale("C")` and `std::locale("")`, but apparently the latter cannot always be relied upon. See: https://stackoverflow.com/a/46364980

The original was failing in our AWS docker image (but passing in our docker-based CI, and everywhere else I've run it).

See also: https://github.com/LLNL/axom/issues/1128

File: src/examples/radiuss_tutorial/lesson_03/README.md [hunk 15]
```diff
@@ -0,0 +1,358 @@
+# Lesson 03: Naive self-intersections... now on the device!
+
+In this lesson, we will port our naive self-intersection algorithm from the previous lesson to different execution spaces, including serial (sequential), threaded (OpenMP) and/or GPU-based (cuda) backends with the help of the RAJA and Umpire libraries. These libraries allow a single implementation to run on different execution spaces.
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_03/README.md:3]:
```suggestion
In this lesson, we will port our naive self-intersection algorithm from the previous lesson to different execution spaces, including serial (sequential), threaded (OpenMP) and/or GPU-based (CUDA) backends with the help of the RAJA and Umpire libraries. These libraries allow a single implementation to run on different execution spaces.
```

File: src/examples/radiuss_tutorial/lesson_03/README.md [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+We also add user support via the following ``RuntimePolicy`` enum:
+```cpp
+enum class RuntimePolicy
+{
+  raja_seq = 1,
+  raja_omp = 2,
+  raja_cuda = 3
+};
+```
+and add a ``RuntimePolicy`` member to  our ``Input`` class
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_03/README.md:57]:
```suggestion
and add a ``RuntimePolicy`` member to  our ``Input`` class:
```

File: src/examples/radiuss_tutorial/lesson_04/device_spatial_indexes.cpp [hunk 17]
```diff
@@ -0,0 +1,582 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+//-----------------------------------------------------------------------------
+///
+/// file: device_self_intersections.cpp
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_04/device_spatial_indexes.cpp:8]:
```suggestion
/// file: device_spatial_indexes.cpp
```

File: src/examples/radiuss_tutorial/lesson_04/device_spatial_indexes.cpp [hunk 18]
```diff
@@ -0,0 +1,582 @@
...
+///
+/// This example ports the naive self-intersection algorithm to additional
+/// execution and memory spaces w/ the help of RAJA and Umpire
```
Comment (white238) [inline src/examples/radiuss_tutorial/lesson_04/device_spatial_indexes.cpp:11]: Check this description since there was a copy/paste error above.

**Review by white238 — APPROVED** [review]
Thanks for the hard work Kenny! This looks great and I look forward for us to continue to build on it.

File: src/examples/radiuss_tutorial/host-config.cmake.in (hunk 6 above)
Comment (kennyweiss) [reply src/examples/radiuss_tutorial/host-config.cmake.in:13]: No -- I adapted this from the `using_with_blt` one. I'm pretty sure our TPLs were also built w/ fortran in the docker image

File: src/examples/radiuss_tutorial/lesson_00/README.md (hunk 1 above)
Comment (kennyweiss) [reply src/examples/radiuss_tutorial/lesson_00/README.md]: I switched the alerts to things that can be rendered in github

File: src/examples/radiuss_tutorial/lesson_04/device_spatial_indexes.cpp (hunk 18 above)
Comment (kennyweiss) [reply src/examples/radiuss_tutorial/lesson_04/device_spatial_indexes.cpp:11]: Thanks! I forgot to update it!

Comment (kennyweiss) [reply src/examples/radiuss_tutorial/lesson_04/device_spatial_indexes.cpp]: Done!

**Comment by kennyweiss** [comment]
Docs build for this branch w/ working graphviz: https://axom.readthedocs.io/en/feature-kweiss-aws-tutorial/

File: src/examples/radiuss_tutorial/lesson_03/device_self_intersections.cpp [hunk 19]
```diff
@@ -0,0 +1,550 @@
...
+
+  // Phase II: If there are intersections, we need to extract their indices to an array
+  if(numIntersect.get() > 0)
+  {
+    const auto numIndices = 2 * numIntersect.get();
+
+    // allocate an array for the intersections (without initializing it)
+    IndexArray intersections_d(axom::ArrayOptions::Uninitialized {},
+                               numIndices,
+                               kernel_allocator);
```
Comment (bmhan12) [inline src/examples/radiuss_tutorial/lesson_03/device_self_intersections.cpp:429]:
```suggestion
    IndexArray intersections_d(axom::ArrayOptions::Uninitialized {},
                               numIndices,
                               numIndices,
                               kernel_allocator);
```

Bugfix for CUDA policy. Lesson 4 also has a few instances of this constructor call as well. Looks like the uninitialized constructor does work, just uses a different overload. @kennyweiss reported this works on rzansel as-is, maybe it's the Automatic Translatation Service covering for us?

Comment (kennyweiss) [reply src/examples/radiuss_tutorial/lesson_03/device_self_intersections.cpp:429]: Thanks @bmhan12

My solution was to use the initializing constructor -- e,g,
```cpp
    IndexArray intersections_d(numIndices,numIndices,kernel_allocator);
```
Yours is better!

I'll update the code for `lesson_03` and `lesson_04`.

File: src/examples/radiuss_tutorial/lesson_01/README.md (hunk 11 above)
Comment (kennyweiss) [reply src/examples/radiuss_tutorial/lesson_01/README.md:51]: Reworded an capitalized.

File: src/axom/primal/geometry/BoundingBox.hpp (hunk 2 above)
Comment (kennyweiss) [reply src/axom/primal/geometry/BoundingBox.hpp:456]: I kept the updated bbox intersect changes.
We can revisit if we need to use-platform specific overrides in a follow-up PR.

This is tracked in https://github.com/LLNL/axom/issues/582

File: scripts/llnl_scripts/llnl_lc_build_tools.py [hunk 20]
```diff
@@ -378,6 +379,34 @@ def build_and_test_host_config(test_root, host_config,
             print("[ERROR: Installed 'using-with-blt' example for host-config: %s failed]\n\n" % host_config)
             return res
 
+    if should_test_installed_tutorial:
```
Comment (kennyweiss) [inline scripts/llnl_scripts/llnl_lc_build_tools.py:382]: @white238 -- please double-check this change to our `build_src` script

File: src/examples/radiuss_tutorial/README.md [hunk 21]
```diff
@@ -0,0 +1,16 @@
+## Axom tutorial
```
Comment (kennyweiss) [inline src/examples/radiuss_tutorial/README.md:1]: Note: New high-level `README` w/ overview of the tutorial

File: scripts/llnl_scripts/llnl_lc_build_tools.py [hunk 22]
```diff
@@ -378,6 +379,34 @@ def build_and_test_host_config(test_root, host_config,
...
+
+        example_commands = [
+            "cd {0}".format(install_example_dir),
+            "rm -rf build",
+            "mkdir build",
+            "cd build",
+            """echo "[Configuring 'radiuss_tutorial']" """,
+            "cmake -C ../host-config.cmake -DCMAKE_BUILD_TYPE={0} ..".format(build_type),
+            """echo "[Building 'radiuss_tutorial']" """,
+            "make -j16",
```
Comment (white238) [inline scripts/llnl_scripts/llnl_lc_build_tools.py:397]: 16 may be generous here and on the ctest command.. I would go with 8 at most.

Comment (kennyweiss) [reply scripts/llnl_scripts/llnl_lc_build_tools.py:397]: Sure. We use `-j16` for building in a bunch of places.
Are you suggestion -`j8` for ctest, but `-j16` for building?

**Review by bmhan12 — APPROVED** [review]

File: scripts/llnl_scripts/llnl_lc_build_tools.py (hunk 22 above)
Comment (white238) [reply scripts/llnl_scripts/llnl_lc_build_tools.py:397]: I must be mistaking this for Serac... we ran into a problem and had to lower the job parallism. You are right we have 16 all over. Its fine to just leave it then.

**Pull request merged**